### PR TITLE
Fixes #6040 Preload mobile homepage on activation & RUCSS clean

### LINKED ISSUE #6040: The mobile version of the homepage isn't preloaded in priority when separate cache for mobile devices is enabled
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version :heavy_check_mark: `3.14.1`
- Used the search feature to ensure that the bug hasn’t been reported before :heavy_check_mark: 

**Describe the bug**

On certain occasions, e.g. when activating WP Rocket, or when clearing the used CSS, we are preloading the homepage in priority:
https://github.com/wp-media/wp-rocket/blob/4c697c3ff0487337a1ede10946d4b2a97b41395d/inc/Engine/Activation/Activation.php#L94-L102

If the "Separate cached files for mobile devices" is enabled we aren't preloading that version. Eventually, the mobile-specific file will be preloaded but that can take a while.

**To Reproduce**

1. Enable "Separate cached files for mobile devices" and Optimize CSS Delivery.
2. Clear the used CSS.
3. Look into the `/cache/wp-rocket/` folder.
4. The desktop version of the homepage is created immediately while for the mobile version, it takes a while.

**Expected behavior**

Preload the mobile-specific cached file of the homepage whenever we do that for the desktop version.

**Screenshots**

N/A

**Additional context**

The `user-agent` we use is `WP Rocket/Homepage Preload`. We need to add a request using a mobile `user-agent` wherever that one is used.

**Backlog Grooming (for WP Media dev team use only)**
- [ ] Reproduce the problem
- [ ] Identify the root cause
- [ ] Scope a solution
- [ ] Estimate the effort

### PR DESCRIPTION
**Pull request by remyperona**
## Description

Preload mobile homepage on activation, and fix bug where preload happened before cache clean when doing a RUCSS clean.

Fixes #6040

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Is the solution different from the one proposed during the grooming?

No

## How Has This Been Tested?

- [x] Automated tests

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by mostafa-hisham — APPROVED** [review]
Looks good to me 
@wp-media/php 
but better have another quick look at it  
I didn't work a lot in the preload part

File: inc/Engine/Preload/ServiceProvider.php [hunk 1]
```diff
@@ -63,102 +60,103 @@ class ServiceProvider extends AbstractServiceProvider {
 	 * @return void
 	 */
 	public function register() {
-		$options = $this->getContainer()->get( 'options' );
+		$options = $this->getLeagueContainer()->get( 'options' );
```
Comment (engahmeds3ed) [inline inc/Engine/Preload/ServiceProvider.php]: really I don't know the exact difference betwee both methods to get the container but as in here:
https://container.thephpleague.com/4.x/upgrade-guide/

> The distinction between AbstractServiceProvider::getContainer and AbstractServiceProvider::getLeagueContainer has been removed. Always just use AbstractServiceProvider::getContainer

**Review by engahmeds3ed — CHANGES_REQUESTED** [review]

**Review by engahmeds3ed — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@Tabrisrp thanks for the PR. please note that currently, after a fresh install of wpr , the cache of the home page is not added to the cache folder also it is completed in the cache table. can you please check that :pray:

File: inc/Engine/Preload/ServiceProvider.php (hunk 1 above)
Comment (CrochetFeve0251) [reply inc/Engine/Preload/ServiceProvider.php]: @engahmeds3ed the difference is that `getContainer` returns the `ContainerInterface` interface from the PSR and `getLeagueContainer` returns the `Container` from League library

**Comment by remyperona** [comment]
From my local testing:

It seems there is a timing issue at play here, when running the activation process, we create or modify multiple files (wp-config, advanced-cache, cache config), and WP Rocket itself is not fully loaded (confirmed by using the debug log on the cache process), so the preload requests to the homepage are not doing anything.

We would have to find a different solution for this point, as I expect it was not working correctly before either anyway.

**Comment by piotrbak** [comment]
It's not a regression then, right @Mai-Saad ?

**Comment by Mai-Saad** [comment]
Yes @piotrbak. The following is happening already on 3.14.2.1
- on trunk randomly (often), after fresh activation, the home URL is completed at the cache table but not in the cache folder (with custom sitemap or WordPress sitemap)
- on trunk randomly (rarely), after fresh install and activation of wpr then we enable separate mobile cache => only desktop is at cache folder although status is completed
**Note:** 
- when using 3.12.0.5, usually it works fine regarding the home page added to the cache folder once completed, rarely (with a bad internet connection) after fresh activation of WPR, the home page is added as completed in the database (without adding a file in the cache folder yet) then the sitemap is fetched and the URL  status change to pending so the next processing URL is added at cache folder.
- 3rd point of AC won't work with fresh activation

**Review by Mai-Saad — APPROVED** [review]
Now both mobile/desktop is added to the used CSS after clearing the used CSS and activating WPR while a separate mobile cache is on will add both mobile and desktop
 **note:**
- fresh activation won't add a cache file in the cache folder although the status is completed as same as with 3.14.2.1
- failed tests are also failing on 3.14.2.1
- if RUCSS and separate mobile cache are enabled at the same instant, we create the used CSS for the desktop of the home then preload the home then create the used CSS for mobile then preload home (same on 3.14.2.1)
- clear and preload cache while separate + rucss on, visit url while home and that URL are still pending=> sometimes loads only desktop (same on 3.14.2.1)
- once on the branch when disabling separate mobile cache if RUCSS was on and completed  => still have a mobile  version in the cache folder (same sometimes on 3.14.2.1) (if after that we enabled separate mobile cache => only desktop added)
- activate while
1- cache is completed at the table but the folder is empty => adds the files for home (often, mainly when with activation or refresh admin after activation, other URLs added to the table and we here add home)
2- cache folder empty, cache table completed => adds files for home (other URLs added to cache table)
[testrail-report-509.pdf](https://github.com/wp-media/wp-rocket/files/12229425/testrail-report-509.pdf)

**Pull request merged**